File: cowork-roadmap/src/main/java/com/cowork/roadmap/global/team/TeamMemberEventConsumer.java
package com.cowork.roadmap.global.team;

import java.time.Duration;

import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Component;

import lombok.RequiredArgsConstructor;
import tools.jackson.databind.ObjectMapper;

@Component
@RequiredArgsConstructor
public class TeamMemberEventConsumer {

    private static final Logger LOG = LoggerFactory.getLogger(TeamMemberEventConsumer.class);
    private static final Duration APPLY_TIMEOUT = Duration.ofSeconds(10);

    private final ObjectMapper objectMapper;
    private final TeamMemberProjectionProcessor processor;

    @KafkaListener(topics = "${KAFKA_TOPIC_TEAM_MEMBER:team.member.event}", groupId = "${KAFKA_GROUP_ID_TEAM_MEMBER:cowork-roadmap.team-member}", containerFactory = "teamMemberListenerContainerFactory")
    public void consume(ConsumerRecord<String, String> record) {
        if (ProjectionSnapshotCompletion.isReserved(record)) {
            String violation = ProjectionSnapshotCompletion.violation(objectMapper, record);
            if (violation == null) {
                processor.completeSnapshot(record, ProjectionSnapshotCompletion.snapshotId(objectMapper, record))
                        .block(APPLY_TIMEOUT);
            } else {
                LOG.error("Discarding invalid projection snapshot marker: {}", violation);
                discard(record, violation);
            }
            return;
        }

        final TeamMemberEvent event;
        try {
            event = objectMapper.readValue(record.value(), TeamMemberEvent.class);
        } catch (Exception exception) {
            LOG.error("Discarding malformed team.member.event payload", exception);
            discard(record, "team.member.event JSON is malformed: " + exception.getMessage());
            return;
        }
        if (event == null) {
            discard(record, "team.member.event top-level null is not allowed");
            return;
        }

        String expectedKey = event.teamId() + ":" + event.userId();
        if (!expectedKey.equals(record.key())) {
            LOG.error("Discarding team.member.event with mismatched key: expected={}, actual={}",
                    expectedKey,
                    record.key());
            discard(record, "team.member.event key does not match payload identity");
            return;
        }

        try {
            processor.apply(record, event).block(APPLY_TIMEOUT);
        } catch (IllegalArgumentException exception) {
            LOG.error("Discarding invalid team.member.event payload", exception);
            discard(record, exception.getMessage());
        }
    }

    private void discard(ConsumerRecord<String, String> record, String reason) {
        processor.discard(record, reason == null ? "invalid team.member.event" : reason).block(APPLY_TIMEOUT);
    }
}

File: cowork-project/src/main/kotlin/com/cowork/project/global/consumer/TeamMemberEventConsumer.kt
package com.cowork.project.global.consumer

import com.cowork.project.global.projection.ProjectionRecordProcessor
import com.cowork.project.global.projection.ProjectionStreams
import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.slf4j.LoggerFactory
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.stereotype.Component

private const val TEAM_MEMBER_EVENT_TOPIC = "team.member.event"

@Component
class TeamMemberEventConsumer(
    private val handler: ProjectLifecycleHandler,
    private val objectMapper: ObjectMapper,
    private val processor: ProjectionRecordProcessor,
    private val streams: ProjectionStreams,
) {
    private val log = LoggerFactory.getLogger(TeamMemberEventConsumer::class.java)

    @KafkaListener(
        topics = [TEAM_MEMBER_EVENT_TOPIC],
        groupId = "cowork-project.team-member",
        containerFactory = "teamMemberEventListenerContainerFactory",
    )
    fun consume(record: ConsumerRecord<String, String>) {
        if (processor.processControlRecord(streams.teamMember, record)) return
        val payload = runCatching {
            requireNotNull(objectMapper.readValue(record.value(), TeamMemberEventPayload::class.java)) {
                "top-level null은 허용되지 않습니다."
            }
        }
            .getOrElse {
                quarantine(record, "team.member.event JSON 역직렬화 실패: ${it.message}")
                return
            }
        val expectedKey = "${payload.teamId}:${payload.userId}"
        val occurredAt = payload.occurredAt
        val reason = when {
            record.key() != expectedKey -> "teamId:userId와 Kafka key가 일치하지 않습니다."
            payload.teamId <= 0 || payload.userId <= 0 -> "teamId와 userId는 양수여야 합니다."
            occurredAt == null -> "occurredAt이 필요합니다."
            payload.role !in setOf("OWNER", "ADMIN", "MEMBER") -> "지원하지 않는 role입니다."
            payload.eventType !in setOf("UPSERT", "DELETE") -> "지원하지 않는 eventType입니다."
            else -> null
        }
        if (reason != null) {
            quarantine(record, reason)
            return
        }
        val eventOccurredAt = requireNotNull(occurredAt)

        processor.applyRecord(streams.teamMember, record) {
            when (payload.eventType) {
                "UPSERT" -> handler.onMemberUpsert(payload.teamId, payload.userId, payload.role, eventOccurredAt)
                "DELETE" -> handler.onMemberRemovedFromTeam(
                    payload.teamId,
                    payload.userId,
                    payload.role,
                    eventOccurredAt,
                )
            }
        }
    }

    private fun quarantine(record: ConsumerRecord<String, String>, reason: String) {
        log.warn(
            "team.member.event를 격리합니다 [partition={}, offset={}, reason={}]",
            record.partition(),
            record.offset(),
            reason,
        )
        processor.quarantineRecord(streams.teamMember, record, reason)
    }
}
